pkg/core/sync.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	sta, err := state.NewMigratedClient(ctx, storage.DSN(), logging.NewZHcLog(&log.Logger, "sync"))
+	if err != nil {
+		return nil, diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer sta.Close()
+
+	cur, err := sta.GetProvider(ctx, rp)
+	if err != nil {
+		return nil, diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	res := &SyncResult{
+		NewVersion: opts.Provider.Version,
+	}
+	if cur != nil {
+		res.OldVersion = cur.Version
+	}
+
+	// TODO run inside TX
+
+	switch {
+	case cur == nil || cur.ParsedVersion == nil: // New install (or older provider)
+		log.Debug().Stringer("provider", opts.Provider).Str("version", opts.Provider.Version).Msg("installing provider schema")
+		res.State = Upgraded
+	case cur.ParsedVersion.Equal(want.ParsedVersion): // Same version
+		res.State = NoChange
+	case cur.ParsedVersion.LessThan(want.ParsedVersion): // Upgrade
+		log.Debug().Stringer("provider", opts.Provider).Str("version", opts.Provider.Version).Msg("upgrading provider schema")
+		res.State = Upgraded
+	case cur.ParsedVersion.GreaterThan(want.ParsedVersion): // Downgrade
+		log.Debug().Stringer("provider", opts.Provider).Str("version", opts.Provider.Version).Msg("downgrading provider schema")
+		res.State = Downgraded
+	default:
+		return nil, diag.FromError(fmt.Errorf("sync: unhandled case"), diag.INTERNAL)
+	}
+
+	if res.State == Upgraded || res.State == Downgraded {
+		if err := dropProvider(ctx, storage.DSN(), rp, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("drop provider failed: %w", err), diag.INTERNAL)
+		}
+		if err := installProvider(ctx, storage.DSN(), db, want, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("install provider failed: %w", err), diag.INTERNAL)
+		}
+	}
+
+	log.Debug().Stringer("provider", opts.Provider).Stringer("state", res.State).Msg("provider sync complete")
+	return res, nil
+}
+
+func Drop(ctx context.Context, storage database.Storage, pm *plugin.Manager, provider registry.Provider) diag.Diagnostics {
+	log.Warn().Stringer("provider", provider).Msg("dropping provider schema")
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: provider})
+	if diags.HasDiags() {
+		return diags
+	}
+	log.Info().Str("provider", provider.Name).Str("version", provider.Version).Msg("dropping provider tables")
+	err := dropProvider(ctx, storage.DSN(), provider, s.ResourceTables)
+	if err != nil {
+		return diag.FromError(err, diag.DATABASE, diag.WithSummary("failed to drop provider"))
+	}
+	return nil
+}
+
+func dropProvider(ctx context.Context, dsn string, provider registry.Provider, resourceTables map[string]*schema.Table) error {
+	dType, _, err := sdkdb.ParseDialectDSN(dsn)
+	if err != nil {
+		return err
+	}
+
+	m, err := migrator.New(logging.NewZHcLog(&log.Logger, "migrator"), dType, nil, dsn, fmt.Sprintf("%s_%s", provider.Source, provider.Name))
```

Comment:
feels like the migrator shouldn't exist anymore, didn't see the code in the SDK, so take this comment with a grain of salt.

Reply:
migrator is now only used for drop provider (and schema_migrations tables)

also it exists because we still have the core schema to worry about.